lib/maths/CQuantileSketch.cc

```diff
@@ -416,93 +414,106 @@ void CQuantileSketch::quantile(EInterpolation interpolation,
-                double nl = m_Knots[l].second;
-                double nr = m_Knots[r].second;
-                LOG_TRACE(<< "xl = " << xl << ", nl = " << nl << ", xr = " << xr
-                          << ", nr = " << nr);
-
-                // Find the points that have been merged with xl and xr.
-                std::size_t ll = (CUniqueIterator(m_Knots, l) - 1).index();
-                std::size_t rr = (CUniqueIterator(m_Knots, r) + 1).index();
-
-                double xm = 0.0, nm = 0.0;
-                switch (m_Interpolation) {
-                case E_Linear:
-                    xm = (nl * xl + nr * xr) / (nl + nr);
-                    nm = nl + nr;
-                    break;
-                case E_PiecewiseConstant:
-                    xm = nl < nr ? xr : (nl > nr ? xl : u01(rng) < 0.5 ? xl : xr);
-                    nm = nl + nr;
-                    break;
-                }
-                for (std::size_t i = ll + 1; i < rr; ++i) {
-                    m_Knots[i].first = xm;
-                    m_Knots[i].second = nm;
-                }
-                LOG_TRACE(<< "merged = "
-                          << core::CContainerPrinter::print(&m_Knots[ll + 1], &m_Knots[rr]));
-                LOG_TRACE(<< "right  = " << core::CContainerPrinter::print(m_Knots[rr]));
+        LOG_TRACE(<< "merge costs = " << core::CContainerPrinter::print(mergeCosts));
 
-                if (ll > 0) {
-                    stale.push_back(ll);
-                }
-                if (rr < m_Knots.size() - 2) {
-                    stale.push_back(rr - 1);
-                }
-                ++merged;
-            } else {
-                CUniqueIterator ll(m_Knots, l);
-                costs[l - 1].first = this->cost(*(ll), *(ll + 1));
-                indexing.push_back(l - 1);
-                std::push_heap(indexing.begin(), indexing.end(), CIndexingGreater(costs));
-                stale.erase(std::remove(stale.begin(), stale.end(), l), stale.end());
+        this->reduce(rng, mergeCosts, mergeCandidates, stale);
+    }
+}
+
+std::size_t CQuantileSketch::target() const {
+    return static_cast<std::size_t>(0.9 * static_cast<double>(m_MaxSize) + 1.0);
+}
+
+void CQuantileSketch::reduce(CPRNG::CXorOShiro128Plus& rng,
+                             TFloatFloatPrVec& mergeCosts,
+                             TSizeVec& mergeCandidates,
+                             TBoolVec& stale) {
+
+    auto mergeCostGreater = [&mergeCosts](std::size_t lhs, std::size_t rhs) {
+        return COrderings::lexicographical_compare(
+            -mergeCosts[lhs].first, mergeCosts[lhs].second,
+            -mergeCosts[rhs].first, mergeCosts[rhs].second);
+    };
+    std::make_heap(mergeCandidates.begin(), mergeCandidates.end(), mergeCostGreater);
+
+    std::size_t merged{this->target()};
+    std::ptrdiff_t numberMergeCandidates{static_cast<std::ptrdiff_t>(m_Knots.size()) - 3};
+    boost::random::uniform_01<double> u01;
```

Comment:
Thanks for raising. There is a `std` alternative I'll switch to. I had kept this because quality of random numbers here is probably not critical and this should be faster. But I subsequently rejigged things to cache random number and this consideration no longer applies.

Reply:
See [f10d5fa](https://github.com/elastic/ml-cpp/pull/881/commits/f10d5faeab8789ba7c630b0e9707217a63de4380). Note I needed to make `min` and `max` functions `constexpr` as a result of this change. I also migrated to `cstdint`: see [3501c38](https://github.com/elastic/ml-cpp/pull/881/commits/3501c38a5a668851102e190c26490edde21a2890).